### ISSUE TITLE
Labels with large absolute values require lots of RAM

### ISSUE THREAD

**Issue by clbarnes** [napari/napari#2620]
## 🐛 Bug

This issue has previously been discussed in #2587 , and its specific cause was elucidated here: https://github.com/napari/napari/issues/2587#issuecomment-825338018 .

As a speed optimisation, for a volume containing label with value `n`, an array of length `n` is created (even if label 0 does not co-exist). Therefore, for large values, a very large array is needed. If the array would not fit into memory, a popup warning is raised; if it could theoretically fit but can't be allocated because the pesky computer is actually using some of its RAM already, napari crashes without grace (killed by the OS, i.e. not catchable exceptions).

This doesn't seem to have been a problem so far so it's not something which needs an urgent fix, especially as the optimisation above gave a significant speed boost, but this should be in the backlog.

## To Reproduce

1. View a label array with a large label.

```python
import numpy as np
import napari

dtype = np.dtype("int32")
big_label = np.iinfo(dtype).max
labels = np.array([[big_label]], dtype)
napari.view_labels(labels)
napari.run()
```

## Environment

```
napari: 0.4.8.dev53+g9b46dd5a
Platform: Linux-5.4.112-1-MANJARO-x86_64-with-glibc2.29
System: Manjaro Linux
Python: 3.8.5 (default, Aug 5 2020, 13:09:57) [GCC 10.1.0]
Qt: 5.14.2
PyQt5: 5.14.2
NumPy: 1.19.1
SciPy: 1.5.2
Dask: 2.25.0
VisPy: 0.6.4

OpenGL:
- GL version: 4.6 (Compatibility Profile) Mesa 21.0.2
- MAX_TEXTURE_SIZE: 16384

Screens:
- screen 1: resolution 1920x1080, scale 1.0
- screen 2: resolution 1920x1080, scale 1.0

Plugins:
- console: 0.0.3
- scikit-image: 0.4.8.dev53+g9b46dd5a
- svg: 0.1.4
```

**Comment by clbarnes** [comment]
In the other issue I suggested modulo-ing the absolute label values when they're looked up but that would introduce undesirable regular cycles into the color map and you'd need to take an extra step to mask out the 0 labels.

**Comment by jni** [comment]
> In the other issue I suggested modulo-ing the absolute label values when they're looked up but that would introduce undesirable regular cycles into the color map and you'd need to take an extra step to mask out the 0 labels.

The long term vision here is to directly pass the labels to the GPU and do hashing (modulo or otherwise) on the GPU. The issue with mod is that it is a surprisingly expensive operation!

```python
In [1]: image = np.random.randint(0, 5, size=(2048, 2048))
In [2]: from napari.utils.colormaps import colormap_utils as cu
In [3]: %timeit cu.low_discrepancy_image(image)
220 ms ± 19.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [4]: %timeit image % 257
386 ms ± 18.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

**Comment by jni** [comment]
Closed by #3308!
